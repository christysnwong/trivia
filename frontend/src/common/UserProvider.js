import React, {useEffect, useState} from "react";
import UserContext from "./UserContext";

const demoUser = {
  id: 1,
  username: "testuser",
  first_name: "Test",
  last_name: "User",
  email: "test@test.com",
  stats: {
    level: 14,
    title: "Pro",
    "quizzes completed": 34,
    points: 3440,
    remainingPts: 60,
    levelPts: 300,
  },
};

const UserProvider = ({
  children,
  currUser = demoUser
}) => {
  
  const [currentUser, setCurrUser] = useState();

  return (
    // <UserContext.Provider value={{ currUser }}>
    <UserContext.Provider value={{ currUser, setCurrUser }}>
      {children}
    </UserContext.Provider>
  );
};

export default UserProvider;
