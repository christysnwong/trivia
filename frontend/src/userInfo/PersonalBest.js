import { useState, useEffect, useContext } from "react";
import TriviaApi from "../api/api";
import UserContext from "../common/UserContext";

import Table from "../userTable/Table";

const PersonalBest = () => {
  const [scores, setScores] = useState([]);
  const { currUser } = useContext(UserContext);

  // fetch personal best scores
  useEffect(() => {
    TriviaApi.getScores(currUser.username).then((score) => {
      setScores(score);
    });

  }, [currUser.username]);

  return (
    <div className="col-10 offset-1 col-lg-8 offset-lg-2 col-xxl-6 offset-xxl-3 col-xxxl-4 offset-xxxl-4">

      <h3 className="mb-5">
        <i className="fa-solid fa-trophy me-2"></i>
        Personal Best Scores
      </h3>

      {!scores.length &&
        "There is no score data in this section. Try out a trivia quiz!"}

      {scores.length > 0 && (
        <>
          <p className="mb-4">
            The following table shows your top score in each category and
            difficulty.
          </p>

          <Table entries={scores} />
        </>
      )}
    </div>
  );
};

export default PersonalBest;
