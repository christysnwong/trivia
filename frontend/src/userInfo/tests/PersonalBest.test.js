import React from "react";
import { render, screen } from "@testing-library/react";
import PersonalBest from "../PersonalBest";
import { MemoryRouter } from "react-router-dom";
import UserProvider from "../../common/UserProvider";

import TriviaApi from "../../api/api";

const { commonBeforeEach, commonAfterAll } = require("./_testCommonUserInfo");

beforeEach(commonBeforeEach);
afterAll(commonAfterAll);


it("renders without crashing", function () {
  render(
    <UserProvider>
      <PersonalBest />
    </UserProvider>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <PersonalBest />
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("checks if results from TriviaApi.getScores are fetched and rendered", async () => {
  jest.spyOn(React, "useEffect").mockImplementation((f) => f());
  jest.spyOn(TriviaApi, "getScores");
  render(
    <UserProvider>
      <PersonalBest />
    </UserProvider>
  );

  expect(TriviaApi.getScores).toHaveBeenCalled();

  expect(await screen.findAllByText(/category/i)).toHaveLength(2);
  expect(await screen.findAllByText(/difficulty/i)).toHaveLength(2);
  expect(await screen.findByText(/points/i)).toBeInTheDocument();
  expect(await screen.findAllByText(70)).toHaveLength(3);
  expect(await screen.findByText(/date/i)).toBeInTheDocument();

  expect(await screen.findAllByText(/entertainment/i)).toHaveLength(3);
  expect(await screen.findAllByText(/film/i)).toHaveLength(2);
  expect(await screen.findAllByText(/books/i)).toHaveLength(1);
});