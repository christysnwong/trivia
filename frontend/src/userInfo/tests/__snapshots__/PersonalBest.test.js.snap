// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="col-10 offset-1 col-lg-8 offset-lg-2 col-xxl-6 offset-xxl-3 col-xxxl-4 offset-xxxl-4"
  >
    <h3
      class="mb-5"
    >
      <i
        class="fa-solid fa-trophy me-2"
      />
      Personal Best Scores
    </h3>
    There is no score data in this section. Try out a trivia quiz!
  </div>
</DocumentFragment>
`;
